File: app/schemas/workspace.py
"""
Pydantic Schemas for Unified Operational Workspace & My Work
Paradox Sports OMS - Phase 1 Workspace Enhancements
"""

from datetime import date, datetime, time
from typing import List, Optional
from uuid import UUID
from pydantic import BaseModel, ConfigDict, Field
from app.models.calendar import ActivityCategory, CalendarPriority
from app.models.communication import DirectivePriority
from app.models.event import EventMemberRole, EventStatus, EventType
from app.models.meeting import MeetingStatus, MeetingType, RSVPStatus
from app.models.task import TaskHealth, TaskPriority, TaskStatus, TaskType


class MyWorkTaskItem(BaseModel):
    id: UUID
    title: str
    description: Optional[str] = None
    vertical_id: UUID
    vertical_name: Optional[str] = None
    task_type: TaskType = TaskType.ROUTINE
    priority: TaskPriority
    status: TaskStatus
    health: TaskHealth
    progress_percentage: int
    deadline: Optional[datetime] = None
    blocker_reason: Optional[str] = None
    assigned_to_id: Optional[UUID] = None
    assigned_to_name: Optional[str] = None
    assigned_to_username: Optional[str] = None
    assigned_by_id: Optional[UUID] = None
    assigned_by_name: Optional[str] = None
    assigned_by_username: Optional[str] = None
    event_id: Optional[UUID] = None
    event_title: Optional[str] = None
    created_at: datetime


class MyWorkDirectiveItem(BaseModel):
    id: UUID
    directive_id: UUID
    title: str
    summary: str
    priority: DirectivePriority
    issued_by_name: Optional[str] = None
    deadline: Optional[datetime] = None
    issued_at: datetime


class MyWorkMeetingItem(BaseModel):
    id: UUID
    title: str
    meeting_type: MeetingType
    meeting_date: date
    start_time: Optional[time] = None
    end_time: Optional[time] = None
    location: Optional[str] = None
    meeting_url: Optional[str] = None
    rsvp_status: RSVPStatus
    organizer_name: Optional[str] = None


class MyWorkEventDutyItem(BaseModel):
    event_id: UUID
    title: str
    event_type: EventType
    event_status: EventStatus
    planned_date: date
    role: EventMemberRole
    location: Optional[str] = None


class MyWorkFormItem(BaseModel):
    id: UUID
    form_id: UUID
    form_title: str
    purpose: str
    category: Optional[str] = "Operational"
    status: str
    deadline: Optional[datetime] = None
    vertical_name: Optional[str] = None
    instructions: Optional[str] = None
    created_at: Optional[datetime] = None


class MyWorkReviewItem(BaseModel):
    id: UUID
    item_type: str  # 'FORM_REVIEW' | 'TRANSFER_APPROVAL'
    title: str
    submitted_by_name: Optional[str] = None
    submitted_at: Optional[datetime] = None
    status: str
    urgency: str = "NORMAL"  # 'NORMAL' | 'HIGH' | 'CRITICAL'
    target_entity_id: Optional[UUID] = None
    link: str


class MyWorkIssueItem(BaseModel):
    id: UUID
    title: str
    status: str
    sensitivity: str
    vertical_name: Optional[str] = None
    event_reference: Optional[str] = None
    raised_by_name: Optional[str] = None
    assigned_to_name: Optional[str] = None
    deadline: Optional[datetime] = None
    escalation_target: Optional[str] = None
    action_required: Optional[str] = None
    created_at: datetime


class MyWorkPriorityItem(BaseModel):
    id: str
    item_type: str  # 'TASK' | 'ISSUE' | 'FORM' | 'REVIEW' | 'APPROVAL'
    title: str
    urgency: str  # 'OVERDUE' | 'CRITICAL' | 'APPROVAL_NEEDED' | 'DEADLINE_SOON' | 'ACTION_REQUIRED'
    urgency_label: str
    due_date: Optional[datetime] = None
    detail: Optional[str] = None
    action_link: str
    action_label: str


class MyWorkUserContext(BaseModel):
    primary_role: str
    operational_level: Optional[int] = None
    responsibilities: List[str] = Field(default_factory=list)
    verticals: List[str] = Field(default_factory=list)
    event_team_profile: Optional[dict] = None
    attention_summary: str = ""
    requires_immediate_attention: bool = False


class MyWorkStats(BaseModel):
    active_tasks: int
    completed_tasks: int = 0
    created_by_me_tasks: int = 0
    pending_directives: int = 0
    upcoming_meetings: int = 0
    event_duties: int = 0
    blocked_tasks: int = 0
    overdue_tasks: int = 0
    active_issues: int = 0
    pending_forms: int = 0
    pending_reviews: int = 0
    pending_approvals: int = 0


class UnifiedMyWorkResponse(BaseModel):
    user_id: UUID
    username: str
    full_name: str
    context: Optional[MyWorkUserContext] = None
    stats: MyWorkStats
    priority_queue: List[MyWorkPriorityItem] = Field(default_factory=list)
    tasks: List[MyWorkTaskItem] = Field(default_factory=list)
    completed_tasks: List[MyWorkTaskItem] = Field(default_factory=list)
    created_by_me_tasks: List[MyWorkTaskItem] = Field(default_factory=list)
    pending_forms: List[MyWorkFormItem] = Field(default_factory=list)
    pending_reviews: List[MyWorkReviewItem] = Field(default_factory=list)
    active_issues: List[MyWorkIssueItem] = Field(default_factory=list)
    pending_directives: List[MyWorkDirectiveItem] = Field(default_factory=list)
    meetings: List[MyWorkMeetingItem] = Field(default_factory=list)
    event_duties: List[MyWorkEventDutyItem] = Field(default_factory=list)
    blockers: List[MyWorkTaskItem] = Field(default_factory=list)
    overdue: List[MyWorkTaskItem] = Field(default_factory=list)

File: frontend/types/workspace.ts
/**
 * Operational Workspace Types
 * Aligned with backend GET /api/v1/workspace/my-work contract.
 */

export type TaskPriority = 'LOW' | 'MEDIUM' | 'HIGH' | 'CRITICAL';
export type TaskStatus = 'NOT_STARTED' | 'TODO' | 'IN_PROGRESS' | 'BLOCKED' | 'COMPLETED' | 'CANCELLED';
export type TaskHealth = 'ON_TRACK' | 'AT_RISK' | 'OVERDUE' | 'BLOCKED' | 'COMPLETE';
export type TaskType = 'ROUTINE' | 'EVENT' | 'MILESTONE' | 'DOCUMENTATION' | 'MEETING_FOLLOW_UP';

export type DirectivePriority = 'STANDARD' | 'URGENT' | 'CRITICAL';
export type MeetingType = 'TEAM_SYNC' | 'EXECUTIVE' | 'PLANNING' | 'EMERGENCY';
export type RSVPStatus = 'INVITED' | 'ACCEPTED' | 'DECLINED' | 'TENTATIVE';
export type EventType = 'TOURNAMENT' | 'TRAINING_CAMP' | 'WORKSHOP' | 'COMPETITION';
export type EventStatus = 'DRAFT' | 'PLANNED' | 'ACTIVE' | 'COMPLETED' | 'CANCELLED';
export type EventMemberRole = 'LEAD' | 'LOGISTICS' | 'VOLUNTEER' | 'MEDIC' | 'MEDIA';

export interface MyWorkTaskItem {
  id: string;
  title: string;
  description?: string;
  vertical_id: string;
  vertical_name?: string;
  task_type?: TaskType;
  priority: TaskPriority;
  status: TaskStatus;
  health: TaskHealth;
  progress_percentage: number;
  deadline?: string;
  blocker_reason?: string;
  assigned_to_id?: string;
  assigned_to_name?: string;
  assigned_to_username?: string;
  assigned_by_id?: string;
  assigned_by_name?: string;
  assigned_by_username?: string;
  event_id?: string;
  event_title?: string;
  created_at: string;
}

export interface MyWorkDirectiveItem {
  id: string;
  directive_id: string;
  title: string;
  summary: string;
  priority: DirectivePriority;
  issued_by_name?: string;
  deadline?: string;
  issued_at: string;
}

export interface MyWorkMeetingItem {
  id: string;
  title: string;
  meeting_type: MeetingType;
  meeting_date: string;
  start_time?: string;
  end_time?: string;
  location?: string;
  meeting_url?: string;
  rsvp_status: RSVPStatus;
  organizer_name?: string;
}

export interface MyWorkEventDutyItem {
  event_id: string;
  title: string;
  event_type: EventType;
  event_status: EventStatus;
  planned_date: string;
  role: EventMemberRole;
  location?: string;
}

export interface MyWorkFormItem {
  id: string;
  form_id: string;
  form_title: string;
  purpose: string;
  category?: string;
  status: string;
  deadline?: string;
  vertical_name?: string;
  instructions?: string;
  created_at?: string;
}

export interface MyWorkReviewItem {
  id: string;
  item_type: 'FORM_REVIEW' | 'TRANSFER_APPROVAL' | string;
  title: string;
  submitted_by_name?: string;
  submitted_at?: string;
  status: string;
  urgency: 'NORMAL' | 'HIGH' | 'CRITICAL' | string;
  target_entity_id?: string;
  link: string;
}

export interface MyWorkIssueItem {
  id: string;
  title: string;
  status: string;
  sensitivity: string;
  vertical_name?: string;
  event_reference?: string;
  raised_by_name?: string;
  assigned_to_name?: string;
  deadline?: string;
  escalation_target?: string;
  action_required?: string;
  created_at: string;
}

export interface MyWorkPriorityItem {
  id: string;
  item_type: 'TASK' | 'ISSUE' | 'FORM' | 'REVIEW' | 'APPROVAL' | string;
  title: string;
  urgency: 'OVERDUE' | 'CRITICAL' | 'APPROVAL_NEEDED' | 'DEADLINE_SOON' | 'ACTION_REQUIRED' | string;
  urgency_label: string;
  due_date?: string;
  detail?: string;
  action_link: string;
  action_label: string;
}

export interface MyWorkEventTeamProfile {
  team_name: string;
  head_name?: string;
  head_email?: string;
  head_phone?: string;
  event_id?: string;
  event_name?: string;
  members_count: number;
}

export interface MyWorkUserContext {
  primary_role: string;
  operational_level?: number;
  responsibilities: string[];
  verticals: string[];
  event_team_profile?: MyWorkEventTeamProfile;
  attention_summary: string;
  requires_immediate_attention: boolean;
}

export interface MyWorkStats {
  active_tasks: number;
  completed_tasks: number;
  created_by_me_tasks: number;
  pending_directives: number;
  upcoming_meetings: number;
  event_duties: number;
  blocked_tasks: number;
  overdue_tasks: number;
  active_issues?: number;
  pending_forms?: number;
  pending_reviews?: number;
  pending_approvals?: number;
}

export interface UnifiedMyWorkResponse {
  user_id: string;
  username: string;
  full_name: string;
  context?: MyWorkUserContext;
  stats: MyWorkStats;
  priority_queue?: MyWorkPriorityItem[];
  tasks: MyWorkTaskItem[];
  completed_tasks: MyWorkTaskItem[];
  created_by_me_tasks: MyWorkTaskItem[];
  pending_forms?: MyWorkFormItem[];
  pending_reviews?: MyWorkReviewItem[];
  active_issues?: MyWorkIssueItem[];
  pending_directives: MyWorkDirectiveItem[];
  meetings: MyWorkMeetingItem[];
  event_duties: MyWorkEventDutyItem[];
  blockers: MyWorkTaskItem[];
  overdue: MyWorkTaskItem[];
}
